src/geom/util/GeometryFixer.cpp

```diff
@@ -0,0 +1,337 @@
+/**********************************************************************
+ *
+ * GEOS - Geometry Engine Open Source
+ * http://geos.osgeo.org
+ *
+ * Copyright (C) 2021 Paul Ramsey <pramsey@cleverelephant.ca>
+ *
+ * This is free software; you can redistribute and/or modify it under
+ * the terms of the GNU Lesser General Public Licence as published
+ * by the Free Software Foundation.
+ * See the COPYING file for more information.
+ *
+ **********************************************************************/
+
+#include <geos/geom/util/GeometryFixer.h>
+#include <geos/geom/Geometry.h>
+#include <geos/geom/GeometryFactory.h>
+#include <geos/operation/overlayng/OverlayNG.h>
+#include <geos/operation/overlayng/OverlayNGRobust.h>
+#include <geos/operation/buffer/BufferOp.h>
+#include <geos/operation/valid/RepeatedPointRemover.h>
+#include <geos/util/UnsupportedOperationException.h>
+
+using geos::operation::valid::RepeatedPointRemover;
+
+namespace geos {
+namespace geom { // geos.geom
+namespace util { // geos.geom.util
+
+/* public static */
+std::unique_ptr<Geometry>
+GeometryFixer::fix(const Geometry* p_geom)
+{
+    GeometryFixer fix(p_geom);
+    return fix.getResult();
+}
+
+/* public */
+void
+GeometryFixer::setKeepCollapsed(bool p_isKeepCollapsed)
+{
+    isKeepCollapsed = p_isKeepCollapsed;
+}
+
+/* public */
+std::unique_ptr<Geometry>
+GeometryFixer::getResult()
+{
+    /**
+     *  Truly empty geometries are simply copied.
+     *  Geometry collections with elements are evaluated on a per-element basis.
+     */
+    if (geom->getNumGeometries() == 0) {
+        return geom->clone();
+    }
+
+    GeometryTypeId typeId = geom->getGeometryTypeId();
+
+    if (typeId == GEOS_POINT)              return fixPoint(static_cast<const Point*>(geom));
```

Comment:
using a switch (typeId ) could save you the final throw that will never be hit, and get a warning the day GeometryTypeId is extended with curve geometries :-)